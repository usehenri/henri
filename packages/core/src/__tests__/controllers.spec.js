const BaseModule = require('../base/module');
const Henri = require('../henri');
const Controllers = require('../2.controllers');

describe('controllers', () => {
  beforeAll(async () => {
    this.henri = new Henri({ runlevel: 2 });
    await this.henri.init();
  });

  afterAll(async () => {
    await this.henri.stop();
  });

  test('should be defined', () => {
    expect(this.henri.controllers).toBeDefined();
  });

  test('should extend BaseModule', () => {
    expect(this.henri.controllers).toBeInstanceOf(BaseModule);
  });

  test('should match snapshot', () => {
    const controllers = new Controllers();
    expect(controllers).toMatchSnapshot();
  });

  test('should load controllers and expose them', async () => {
    const mock = jest.fn();
    const controllers = {
      'someFolder/index': {
        index: () => mock,
        create: () => mock,
      },
      other: {
        update: () => mock,
        badStuff: 'really bad stuff',
      },
    };
    await this.henri.controllers.configure(controllers);
    expect(this.henri.controllers.get('someFolder/index#index')).toBeTruthy();
    expect(this.henri.controllers.get('someFolder/index#create')).toBeTruthy();
    expect(this.henri.controllers.get('other#update')).toBeTruthy();
    expect(this.henri.controllers.get('other#badStuff')).toBeUndefined();
  });

  test('should reload', () => {
    expect(this.henri.controllers.reload()).toBeTruthy();
  });

  test('should have get/set', () => {
    expect(this.henri.controllers.set('some#stuff', () => 'abc')).toBeTruthy();
    expect(this.henri.controllers.get('some#stuff')).toBeTruthy();

    expect(this.henri.controllers.set('some/stuff', 'abc')).toBeFalsy();
    expect(this.henri.controllers.get('some/stuff')).toBeFalsy();
  });

  xtest('should load filesystem structure');
  xtest('possibly freeze the private properties (this._controllers)');
});
