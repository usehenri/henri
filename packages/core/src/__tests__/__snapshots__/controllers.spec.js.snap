// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controllers should match snapshot 1`] = `
Controllers {
  "_controllers": Map {},
  "configure": [Function],
  "get": [Function],
  "henri": null,
  "init": [Function],
  "key": null,
  "load": [Function],
  "name": "controllers",
  "reload": [Function],
  "reloadable": true,
  "runlevel": 2,
  "set": [Function],
}
`;
